server/src/main/java/org/opensearch/common/cache/tier/TieredSpilloverCache.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.tier;
+
+import org.opensearch.common.cache.LoadAwareCacheLoader;
+import org.opensearch.common.cache.RemovalReason;
+import org.opensearch.common.cache.store.Cache;
+import org.opensearch.common.cache.store.StoreAwareCache;
+import org.opensearch.common.cache.store.StoreAwareCacheRemovalNotification;
+import org.opensearch.common.cache.store.StoreAwareCacheValue;
+import org.opensearch.common.cache.store.builders.StoreAwareCacheBuilder;
+import org.opensearch.common.cache.store.enums.CacheStoreType;
+import org.opensearch.common.cache.store.listeners.EventType;
+import org.opensearch.common.cache.store.listeners.StoreAwareCacheEventListener;
+import org.opensearch.common.cache.store.listeners.StoreAwareCacheEventListenerConfiguration;
+import org.opensearch.common.cache.store.listeners.dispatchers.StoreAwareCacheListenerDispatcherDefaultImpl;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * This cache spillover the evicted items from upper tier to lower tier. For now, we are spilling the in-memory
+ * cache items to disk tier cache. All the new items are cached onHeap and if any items are evicted are moved to disk
+ * cache.
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public class TieredSpilloverCache<K, V> implements TieredCache<K, V>, StoreAwareCacheEventListener<K, V> {
+
+    private final Optional<StoreAwareCache<K, V>> onDiskCache;
+    private final StoreAwareCache<K, V> onHeapCache;
```

Comment:
Does it make sense to use TieredCache then if someone wants to use onHeap or onDisk only? 
I think TieredCache in general implies you are using a combination of multiple tiers.
In case someone wants to use onHeap or onDisk only, they can directly use StoreAwareCache?

Let me know what you think.

Reply:
>Good point. But then why DiskCache is optional ? If we plan to have TieredCache as composition of multiple caches, then still we need a mechanism for users to use a) only OnHeap, b) OnDisk and c) TieredCache. Based on this configuration, it will instantiate corresponding cache in IndicesRequestCache

Yeah making DiskCache as optional doesn't make sense then. Should we just keep TieredCache as a composition of onheap and onDisk? Having those different configurations just complicates TieredCache logic and as I mentioned one can directly integrate StoreAwareCache in ```IndicesRequestCache``` if only one is needed.  

I can remove Optional disk cache later when I actually integrate DiskCache (after onHeap) as part of separate PR.

Reply:
> Also we need to make sure TieredSpilloverCache is thread safe as well. I don't see any locking related constructs so far, are we planning to add it later ? Current way of accessing onheap + onDisk cache is not thread safe.

I don't see thread safety issue though with assumption that individual cache tier should be thread safe. The get() method seems thread safe to me for example. Maybe an example would help here.

Reply:
I have added read/write locks. Figured out issue where a key can end up in both onHeap and onDisk cache. Thanks for the catch.